### PR TITLE
HardGLU implementation

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42443 [quant] quantized GLU and HardGLU
* **#42728 HardGLU implementation**
* #42727 Factoring out common code in the GLU
* #42442 [quant] Quantization parameters in TensorIterator

Differential Revision: [D23011805](https://our.internmc.facebook.com/intern/diff/D23011805)

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -1716,11 +1745,11 @@ def silu(input, inplace=False):
         \text{silu}(x) = x * \sigma(x), \text{where } \sigma(x) \text{ is the logistic sigmoid.}
 
     .. note::
-        See `Gaussian Error Linear Units (GELUs) <https://arxiv.org/abs/1606.08415>`_ 
-        where the SiLU (Sigmoid Linear Unit) was originally coined, and see 
-        `Sigmoid-Weighted Linear Units for Neural Network Function Approximation 
-        in Reinforcement Learning <https://arxiv.org/abs/1702.03118>`_ and `Swish: 
-        a Self-Gated Activation Function <https://arxiv.org/abs/1710.05941v1>`_ 
+        See `Gaussian Error Linear Units (GELUs) <https://arxiv.org/abs/1606.08415>`_
```
Comment (z-a-f) [inline torch/nn/functional.py:1748]: Whoops -- auto linter thingy....

File: test/test_torch.py [hunk 2]
```diff
@@ -16581,6 +16581,30 @@ def test_hardsigmoid(self, device, dtype):
                          torch.tensor(expectedOutput, dtype=dtype, device=device),
                          atol=precision_4dps, rtol=0)
 
+    @dtypes(torch.float, torch.double)
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_hardglu(self, device, dtype):
```
Comment (vkuzo) [inline test/test_torch.py:16586]: can we also check the derivative?  `torch.autograd.gradcheck` could be useful

**Review by vkuzo — COMMENTED** [review]
can we add an op benchmark as well?

File: aten/src/ATen/native/cuda/Activation.cu [hunk 3]
```diff
@@ -478,6 +478,36 @@ void hardsigmoid_backward_kernel(TensorIterator& iter) {
   });
 }
 
+void hardglu_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "hardsigmoid_cuda", [&]() {
```
Comment (raghuramank100) [inline aten/src/ATen/native/cuda/Activation.cu:482]: NIT: Typo here "hardsigmoid_cuda"?

File: aten/src/ATen/native/GatedLinearUnit.cpp [hunk 4]
```diff
@@ -86,5 +88,51 @@ Tensor glu_backward(const Tensor& grad_output, const Tensor& input, int64_t dim)
...
+  hardglu_stub(iter.device_type(), iter);
+  return result;
+}
+
+Tensor hardglu(const Tensor& self, int64_t dim) {
+  auto result = at::empty({0}, self.options());
+  return at::hardglu_out(result, self, dim);
+}
+
+Tensor& hardglu_backward_out(Tensor& grad_input,
```
Comment (raghuramank100) [inline aten/src/ATen/native/GatedLinearUnit.cpp:109]: Not sure if I follow the math here. The gradient for hard GLU has two components:
HGLU(x) = x[0:n-1].Hsigmoid([x[n:end]) 
gradient[i] = HS[i+n]  * y[i] if i <n
 else: 
gradient[i+N] = x[0:n-1].Gradient(Hsigmoid). 
Is this what is being done by the code below?
